src/VisualStudio/CSharp/Impl/LanguageService/CSharpCodeCleanupFixer.cs

```diff
@@ -320,37 +340,138 @@ private async Task<bool> FixTextBufferAsync(TextBufferCodeCleanUpScope textBuffe
+                        {
+                            scope.Description = description;
+                            scope.Progress.Report(new ProgressInfo(completed, total));
+                        }
+                    });
+
+                    var document = buffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                    var newDoc = await FixDocumentAsync(document, context.EnabledFixIds, progressTracker, cancellationToken).ConfigureAwait(true);
 
                     await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                    cancellationToken.ThrowIfCancellationRequested();
+
                     return document.Project.Solution.Workspace.TryApplyChanges(newDoc.Project.Solution, progressTracker);
                 }
             }
         }
+
+        private async Task<Solution> FixSolutionAsync(Solution solution, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            progressTracker.AddItems(solution.ProjectIds.Count);
+            foreach (var projectId in solution.ProjectIds)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var project = solution.GetProject(projectId);
+                progressTracker.Description = project.Name;
+                var newProject = await FixProjectAsync(project, enabledFixIds, new ProgressTracker(), cancellationToken).ConfigureAwait(false);
+                solution = newProject.Solution;
+                progressTracker.ItemCompleted();
+            }
+
+            return solution;
+        }
+
+        private async Task<Project> FixProjectAsync(Project project, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            if (project.LanguageServices.GetService<ICodeCleanupService>() is null)
+            {
+                return project;
+            }
+
+            progressTracker.AddItems(project.DocumentIds.Count);
+            foreach (var documentId in project.DocumentIds)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var document = project.GetDocument(documentId);
+                progressTracker.Description = document.Name;
+                var fixedDocument = await FixDocumentAsync(document, enabledFixIds, new ProgressTracker(), cancellationToken).ConfigureAwait(false);
+                project = fixedDocument.Project;
+                progressTracker.ItemCompleted();
+            }
+
+            return project;
+        }
+
+        private async Task<Document> FixDocumentAsync(Document document, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            var codeCleanupService = document.GetLanguageService<ICodeCleanupService>();
+
+            var allDiagnostics = codeCleanupService.GetAllDiagnostics();
```

Comment:
I believe all diagnostics will be same always and enabledFixIds will be same as well for project/solution? shouldn't we create these once rather than doing it repeatedly per document? for something like roslyn which has over 10,000 documents, this array will be generated 10,000 times? or for solution with many multi targetting projects, this might get repeated a lot of times for a solution?

Reply:
I tend to agree with @heejaechang@outlook.com here - this doesn't look like a per-document action.

---
In reply to: [246928624](https://github.com/dotnet/roslyn/pull/32175#discussion_r246928624) [](ancestors = 246928624)

Reply:
➡️ Resolving this as **won't fix** because the suggested refactoring would complicate the code flow for multiple entry points but isn't actually needed for performance.